### PR TITLE
Support formatting specified lines.

### LINKED ISSUE #830: [Feature Request] Allow command line option to only reformat specific lines
Operating system: All
Python version: All Supported
*Black* version: Future
Does also happen on master: Yes

Currently if black is used on an existing project base it will likely create a huge diff as many lines will be changed. Some projects discourage this or only allow it on a major version change.

One way to address this would be to allow per file line ranges to specify that only those lines be considered for re-formatting in a similar manner to `yapf --lines` option. This would allow black reformatting of changed lines only.

This ticket raised in response to https://github.com/python/steering-council/issues/9#issuecomment-489441747 _but of course it could be considered a compromise._

### PR DESCRIPTION
**Pull request by yilei**
### Description

This adds a command-line option `--line-ranges=`, to tell _Black_ to only format these specified lines.

Fixes #830

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

File: docs/usage_and_configuration/the_basics.md [hunk 1]
```diff
@@ -175,6 +175,23 @@ All done! ✨ 🍰 ✨
...
+### `--line-ranges`
+
+When specified, _Black_ will try its best to only format these lines.
+
+This option can be specified multiple times, and a union of the lines will be formatted.
+Each range must be specified as two integers connected by a `-`: `<START>-<END>`. The
+`<START>` and `<END>` integer indices are 1-based and inclusive on both ends.
+
+_Black_ may still format lines outside of the ranges for multi-line statements.
+Formatting more than one files or any ipynb files with this option is not supported.
```
Comment (JelleZijlstra) [inline docs/usage_and_configuration/the_basics.md:187]:
```suggestion
Formatting more than one file or any ipynb files with this option is not supported.
```

File: src/black/ranges.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+        new_range = (new_start, new_end)
+        if is_valid_line_range(new_range):
+            new_lines.append(new_range)
+    return new_lines
+
+
+def convert_unchanged_lines(src_node: Node, lines: Collection[Tuple[int, int]]) -> None:
+    """Converts unchanged lines to STANDALONE_COMMENT.
+
+    The idea is similar to `# fmt: on/off` is implemented. It also converts the
```
Comment (JelleZijlstra) [inline src/black/ranges.py:131]:
```suggestion
    The idea is similar to how `# fmt: on/off` is implemented. It also converts the
```

File: src/black/ranges.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+      -   arg2, arg3):
+      + def func(arg1, arg2, arg3):
+          pass
+    It returns the following mappings:
+      original -> modified
+       (1, 1)  ->  (1, 1), is_changed_block=False (the "import re" line)
+       (2, 3)  ->  (2, 2), is_changed_block=True (the diff)
+       (4, 4)  ->  (3, 3), is_changed_block=False (the "pass" line)
+
+    You can think this visually as if it brings up a side-by-side diff, and tries
```
Comment (JelleZijlstra) [inline src/black/ranges.py:420]:
```suggestion
    You can think of this visually as if it brings up a side-by-side diff, and tries
```

File: src/black/ranges.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+    # Each block corresponds to a _LinesMapping with is_changed_block=False,
+    # and the ranges between two blocks corresponds to a _LinesMapping with
+    # is_changed_block=True,
+    # NOTE: matching_blocks is 0-based, but _LinesMapping is 1-based.
+    for i, block in enumerate(matching_blocks):
+        if i == 0:
+            if block.a != 0 or block.b != 0:
+                lines_mappings.append(
+                    _LinesMapping(
+                        1,
```
Comment (JelleZijlstra) [inline src/black/ranges.py:450]: Keyword arguments would be more clear here.

File: tests/test_ranges.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
+"""Test the ink_adjusted_lines module."""
```
Comment (JelleZijlstra) [inline tests/test_ranges.py:1]:
```suggestion
"""Test the black.ranges module."""
```

**Review by JelleZijlstra — COMMENTED** [review]
Thanks! This looks good, just a few pieces of minor feedback.

This has been a long-standing feature request, so it's great to finally see it in Black.

File: src/black/ranges.py (hunk 4 above)
Comment (yilei) [reply src/black/ranges.py]: Done

**Comment by yilei** [comment]
Thanks for the review! Updated.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by fearedbliss** [comment]
Thanks for implementing this @yilei. This will be extremely useful in many situations and circumstances.

**Comment by henriholopainen** [comment]
Funnily enough, I was working on the same feature and only now realised it had been finished when trying to open a draft PR. 😛 Oh well, what it's worth, my approach can be found [here](https://github.com/henriholopainen/black/tree/issue_830).

**Comment by henriholopainen** [comment]
Also this PR might be enough to close [this issue](https://github.com/psf/black/issues/1352)

**Comment by yilei** [comment]
@henriholopainen Ah I'm sorry I didn't realize you were working on it. I should have sent the PR earlier as I've had it implemented for a while now.

**Comment by tartley** [comment]
This option doesn't appear to be working right for me. I'll post a minimal issue...

**Comment by tartley** [comment]
Minimal repro posted here: https://github.com/psf/black/issues/4033

**Pull request merged**